{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. Prerequisites\n",
    "\n",
    "1. Install git\n",
    "2. clone PEPS\n",
    "3. Install PEPS depenendicies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded plugins: priorities, update-motd, upgrade-helper\n",
      "13 packages excluded due to repository priority protections\n",
      "Resolving Dependencies\n",
      "--> Running transaction check\n",
      "---> Package git.x86_64 0:2.14.6-1.61.amzn1 will be installed\n",
      "--> Processing Dependency: perl-Git = 2.14.6-1.61.amzn1 for package: git-2.14.6-1.61.amzn1.x86_64\n",
      "--> Processing Dependency: perl(Term::ReadKey) for package: git-2.14.6-1.61.amzn1.x86_64\n",
      "--> Processing Dependency: perl(Git::I18N) for package: git-2.14.6-1.61.amzn1.x86_64\n",
      "--> Processing Dependency: perl(Git) for package: git-2.14.6-1.61.amzn1.x86_64\n",
      "--> Processing Dependency: perl(Error) for package: git-2.14.6-1.61.amzn1.x86_64\n",
      "--> Running transaction check\n",
      "---> Package perl-Error.noarch 1:0.17020-2.9.amzn1 will be installed\n",
      "---> Package perl-Git.noarch 0:2.14.6-1.61.amzn1 will be installed\n",
      "---> Package perl-TermReadKey.x86_64 0:2.30-20.9.amzn1 will be installed\n",
      "--> Finished Dependency Resolution\n",
      "\n",
      "Dependencies Resolved\n",
      "\n",
      "================================================================================\n",
      " Package              Arch       Version                 Repository        Size\n",
      "================================================================================\n",
      "Installing:\n",
      " git                  x86_64     2.14.6-1.61.amzn1       amzn-updates      12 M\n",
      "Installing for dependencies:\n",
      " perl-Error           noarch     1:0.17020-2.9.amzn1     amzn-main         33 k\n",
      " perl-Git             noarch     2.14.6-1.61.amzn1       amzn-updates      69 k\n",
      " perl-TermReadKey     x86_64     2.30-20.9.amzn1         amzn-main         33 k\n",
      "\n",
      "Transaction Summary\n",
      "================================================================================\n",
      "Install  1 Package (+3 Dependent packages)\n",
      "\n",
      "Total download size: 12 M\n",
      "Installed size: 29 M\n",
      "Downloading packages:\n",
      "(1/4): perl-Error-0.17020-2.9.amzn1.noarch.rpm             |  33 kB   00:00     \n",
      "(2/4): perl-Git-2.14.6-1.61.amzn1.noarch.rpm               |  69 kB   00:00     \n",
      "(3/4): perl-TermReadKey-2.30-20.9.amzn1.x86_64.rpm         |  33 kB   00:00     \n",
      "(4/4): git-2.14.6-1.61.amzn1.x86_64.rpm                    |  12 MB   00:00     \n",
      "--------------------------------------------------------------------------------\n",
      "Total                                               51 MB/s |  12 MB  00:00     \n",
      "Running transaction check\n",
      "Running transaction test\n",
      "Transaction test succeeded\n",
      "Running transaction\n",
      "  Installing : 1:perl-Error-0.17020-2.9.amzn1.noarch                        1/4 \n",
      "  Installing : perl-TermReadKey-2.30-20.9.amzn1.x86_64                      2/4 \n",
      "  Installing : git-2.14.6-1.61.amzn1.x86_64                                 3/4 \n",
      "  Installing : perl-Git-2.14.6-1.61.amzn1.noarch                            4/4 \n",
      "  Verifying  : 1:perl-Error-0.17020-2.9.amzn1.noarch                        1/4 \n",
      "  Verifying  : git-2.14.6-1.61.amzn1.x86_64                                 2/4 \n",
      "  Verifying  : perl-TermReadKey-2.30-20.9.amzn1.x86_64                      3/4 \n",
      "  Verifying  : perl-Git-2.14.6-1.61.amzn1.noarch                            4/4 \n",
      "\n",
      "Installed:\n",
      "  git.x86_64 0:2.14.6-1.61.amzn1                                                \n",
      "\n",
      "Dependency Installed:\n",
      "  perl-Error.noarch 1:0.17020-2.9.amzn1     perl-Git.noarch 0:2.14.6-1.61.amzn1\n",
      "  perl-TermReadKey.x86_64 0:2.30-20.9.amzn1\n",
      "\n",
      "Complete!\n"
     ]
    }
   ],
   "source": [
    "## Install Git\n",
    "import sys\n",
    "!sudo yum -y install git\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'PEPS'...\n",
      "remote: Enumerating objects: 45, done.\u001b[K\n",
      "remote: Total 45 (delta 0), reused 0 (delta 0), pack-reused 45\u001b[K\n",
      "Unpacking objects: 100% (45/45), done.\n"
     ]
    }
   ],
   "source": [
    "#clone PEPS repo\n",
    "!git clone https://github.com/aehrc/PEPS.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting https://github.com/dceoy/pdbio/archive/master.tar.gz\n",
      "\u001b[?25l  Downloading https://github.com/dceoy/pdbio/archive/master.tar.gz\n",
      "\u001b[K     \\ 20kB 26.7MB/s\n",
      "\u001b[?25hCollecting docopt (from pdbio==0.4.2)\n",
      "  Downloading https://files.pythonhosted.org/packages/a2/55/8f8cab2afd404cf578136ef2cc5dfb50baa1761b68c9da1fb1e4eed343c9/docopt-0.6.2.tar.gz\n",
      "Requirement already satisfied, skipping upgrade: pandas in ./biospark/lib/python3.6/site-packages (from pdbio==0.4.2) (0.23.4)\n",
      "Requirement already satisfied, skipping upgrade: python-dateutil>=2.5.0 in ./biospark/lib/python3.6/site-packages (from pandas->pdbio==0.4.2) (2.8.1)\n",
      "Requirement already satisfied, skipping upgrade: pytz>=2011k in ./biospark/lib/python3.6/site-packages (from pandas->pdbio==0.4.2) (2019.3)\n",
      "Requirement already satisfied, skipping upgrade: numpy>=1.9.0 in ./biospark/lib/python3.6/site-packages (from pandas->pdbio==0.4.2) (1.18.1)\n",
      "Requirement already satisfied, skipping upgrade: six>=1.5 in ./biospark/lib/python3.6/site-packages (from python-dateutil>=2.5.0->pandas->pdbio==0.4.2) (1.14.0)\n",
      "Building wheels for collected packages: pdbio, docopt\n",
      "  Building wheel for pdbio (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /mnt/tmp/pip-ephem-wheel-cache-mdtr4nb7/wheels/e5/5a/c7/0ecdb2f03a7e8c3b384bf01a5e07d62a8dfa8920085abcaa06\n",
      "  Building wheel for docopt (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /home/hadoop/.cache/pip/wheels/9b/04/dd/7daf4150b6d9b12949298737de9431a324d4b797ffd63f526e\n",
      "Successfully built pdbio docopt\n",
      "Installing collected packages: docopt, pdbio\n",
      "Successfully installed docopt-0.6.2 pdbio-0.4.2\n",
      "Collecting scikit-learn\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/d1/48/e9fa9e252abcd1447eff6f9257636af31758a6e46fd5ce5d3c879f6907cb/scikit_learn-0.22.1-cp36-cp36m-manylinux1_x86_64.whl (7.0MB)\n",
      "\u001b[K     |████████████████████████████████| 7.1MB 45.4MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied, skipping upgrade: numpy>=1.11.0 in ./biospark/lib/python3.6/site-packages (from scikit-learn) (1.18.1)\n",
      "Requirement already satisfied, skipping upgrade: scipy>=0.17.0 in ./biospark/lib/python3.6/site-packages (from scikit-learn) (1.3.3)\n",
      "Collecting joblib>=0.11 (from scikit-learn)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/28/5c/cf6a2b65a321c4a209efcdf64c2689efae2cb62661f8f6f4bb28547cf1bf/joblib-0.14.1-py2.py3-none-any.whl (294kB)\n",
      "\u001b[K     |████████████████████████████████| 296kB 51.2MB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: joblib, scikit-learn\n",
      "Successfully installed joblib-0.14.1 scikit-learn-0.22.1\n"
     ]
    }
   ],
   "source": [
    "#Install dependency https://github.com/dceoy/pdbio\n",
    "!{sys.executable} -m pip install -U https://github.com/dceoy/pdbio/archive/master.tar.gz\n",
    "!{sys.executable} -m pip install -U scikit-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obtaining rbo from git+https://github.com/changyaochen/rbo.git@master#egg=rbo\n",
      "  Updating ./src/rbo clone (to revision master)\n",
      "  Running command git fetch -q --tags\n",
      "  Running command git reset --hard -q 2b2ff7b24534f342a5153573bf384c06317ec2de\n",
      "Requirement already satisfied: numpy in ./biospark/lib/python3.6/site-packages (from rbo) (1.18.1)\n",
      "Installing collected packages: rbo\n",
      "  Found existing installation: rbo 0.1\n",
      "    Uninstalling rbo-0.1:\n",
      "      Successfully uninstalled rbo-0.1\n",
      "  Running setup.py develop for rbo\n",
      "Successfully installed rbo\n"
     ]
    }
   ],
   "source": [
    "# https://github.com/changyaochen/rbo\n",
    "!sudo {sys.executable} -m pip install -e git+https://github.com/changyaochen/rbo.git@master#egg=rbo\n",
    "# (might have to install this outside of the jupyter notebook...)\n",
    "# Do I really need this??\n",
    "#!{sys.executable} -m pip install --user pipenv\n",
    "#!{sys.executable} -m pipenv install -e git+https://github.com/changyaochen/rbo.git@master#egg=rbo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Helper scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to create new config file\n",
    "\n",
    "import json\n",
    "import os\n",
    "\n",
    "def newConfigFile(outputPrefix,inputPrefix,outfile,v1,v2,v3,v4,v5):\n",
    "    with open(\"PEPS/SampleData/config-small.json\", 'r') as f:\n",
    "        config = json.load(f)\n",
    "\n",
    "        config['inputType']='csv'\n",
    "        config['outputPrefix']=outputPrefix\n",
    "        config['inputPrefix']=inputPrefix\n",
    "        config['variables'][0]['numVar']=v1\n",
    "        config['variables'][1]['numVar']=v2\n",
    "        config['variables'][2]['numVar']=v3\n",
    "        config['variables'][3]['numVar']=v4\n",
    "        config['variables'][4]['numVar']=v5\n",
    "\n",
    "    os.makedirs(os.path.dirname(outfile), exist_ok=True)\n",
    "    with open(outfile+\".json\", 'w') as outfile:\n",
    "        json.dump(config, outfile, sort_keys=True, indent=4)\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "PEPSINPUT=\"/home/hadoop/PEPS/1000Genomes/all-snps-header\"\n",
    "LOCALDIR=\"/home/hadoop/PEPS/Test\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip $PEPSINPUT\".csv.gz\"\n",
    "# just renaming and unzipping does not seem to work \n",
    "#!mv /home/hadoop/PEPS/1000Genomes/all-snps.vcf.bgz /home/hadoop/PEPS/1000Genomes/all-snps.vcf.gz\n",
    "#!gunzip /home/hadoop/PEPS/1000Genomes/all-snps.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input data \n",
    "#newConfigFile(\"PEPS/SampleData/output-small\",\n",
    "#              \"PEPS/SampleData/input-small\",\n",
    "#              \"PEPS/SampleData/config-small-mod\",\n",
    "#              12,3,3,6,10)\n",
    "\n",
    "#v1=11; v2=3; v3=3; v4=6; v5=10\n",
    "#v1=1; v2=3; v3=3; v4=6; v5=20\n",
    "#v1=20; v2=3; v3=3; v4=6; v5=1\n",
    "#v1=2; v2=4; v3=6; v4=8; v5=10\n",
    "#v1=10; v2=8; v3=6; v4=4; v5=2 <- OK results\n",
    "v1=5; v2=4; v3=3; v4=2; v5=1\n",
    "\n",
    "\n",
    "\"_\"+\"_\".join(str(e) for e in [v1,v2,v3,v4,v5])\n",
    "\n",
    "BASE=\"all-snps-header_\"+\"_\".join(str(e) for e in [v1,v2,v3,v4,v5])\n",
    "\n",
    "newConfigFile(LOCALDIR+\"/\"+BASE,\n",
    "              PEPSINPUT,\n",
    "              LOCALDIR+\"/\"+BASE,\n",
    "              v1,v2,v3,v4,v5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========== Configuration >>>\n",
      "{'dumpCSV': True,\n",
      " 'inputPrefix': '/home/hadoop/PEPS/1000Genomes/all-snps-header',\n",
      " 'inputType': 'csv',\n",
      " 'maxOrder': 5,\n",
      " 'numLoop': 1,\n",
      " 'numSnpsNeeded': 35,\n",
      " 'numTree': 500,\n",
      " 'numVariables': 15,\n",
      " 'outputPrefix': '/home/hadoop/PEPS/Test/all-snps-header_5_4_3_2_1',\n",
      " 'pvalueThr': 0.001,\n",
      " 'shuffleSnps': True,\n",
      " 'variables': [{'numSNPs': 5, 'numSnpsInVar': 1, 'numVar': 5},\n",
      "               {'numSNPs': 8, 'numSnpsInVar': 2, 'numVar': 4},\n",
      "               {'numSNPs': 9, 'numSnpsInVar': 3, 'numVar': 3},\n",
      "               {'numSNPs': 8, 'numSnpsInVar': 4, 'numVar': 2},\n",
      "               {'numSNPs': 5, 'numSnpsInVar': 5, 'numVar': 1}]}\n",
      "============================\n",
      "Number of SNPs in the input file:  17373\n",
      "Number of SNPs needed:  35\n",
      "There are enough SNPs in the input file\n",
      "number of sample 2504\n",
      "number of snp 17373\n",
      "loop 1\n",
      "Done\n",
      "Number of Variables:  15\n",
      "Number of Truth Variables:  9\n",
      "Number of SNPs used to form Variables:  35\n",
      "Number of Truth SNPs (SNPs in Truth Variables):  28\n",
      "PEPS/PEPS.py:465: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  snpDataTruth['lbl'] = varGT['lbl']\n",
      "PEPS/PEPS.py:466: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  snpDataVar['lbl'] = varGT['lbl']\n",
      "AUC Truth SNPs            :  0.7295136884637585\n",
      "AUC All SNPs in Variables :  0.7374404761904763\n",
      "AUC All SNPs in input file:  0.5520164609053497\n"
     ]
    }
   ],
   "source": [
    "!{sys.executable} PEPS/PEPS.py $LOCALDIR/$BASE\".json\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# results\n",
    "\n",
    "Running twice to get a little bit of a feeling for the spread\n",
    "\n",
    "## balanced\n",
    "\n",
    "```\n",
    "v1=10; v2=8; v3=6; v4=4; v5=2\n",
    "AUC Truth SNPs            :  0.6566427686316161\n",
    "AUC All SNPs in Variables :  0.6886058068006493\n",
    "AUC All SNPs in input file:  0.5542574656981435\n",
    "\n",
    "AUC Truth SNPs            :  0.6848957481825414\n",
    "AUC All SNPs in Variables :  0.6316260899594233\n",
    "AUC All SNPs in input file:  0.5071729908633228\n",
    "\n",
    "v1=5; v2=4; v3=3; v4=2; v5=1\n",
    "AUC Truth SNPs            :  0.7295136884637585\n",
    "AUC All SNPs in Variables :  0.7374404761904763\n",
    "AUC All SNPs in input file:  0.5520164609053497\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "```\n",
    "v1=11; v2=3; v3=3; v4=6; v5=10\n",
    "AUC Truth SNPs            :  0.5945935235507246\n",
    "AUC All SNPs in Variables :  0.5879051508157267\n",
    "AUC All SNPs in input file:  0.48481233847087507\n",
    "\n",
    "AUC Truth SNPs            :  0.5866592799049621\n",
    "AUC All SNPs in Variables :  0.5947047619047621\n",
    "AUC All SNPs in input file:  0.5366248137901681\n",
    "```\n",
    "____________________________________________________\n",
    "\n",
    "\n",
    "## weight towards multiple variants\n",
    "\n",
    "```\n",
    "v1=1; v2=3; v3=3; v4=6; v5=20\n",
    "AUC Truth SNPs            :  0.5520690460950756\n",
    "AUC All SNPs in Variables :  0.544824642385618\n",
    "AUC All SNPs in input file:  0.5093466424682396\n",
    "\n",
    "AUC Truth SNPs            :  0.531187248995984\n",
    "AUC All SNPs in Variables :  0.5479636094361579\n",
    "AUC All SNPs in input file:  0.45317957451721846\n",
    "```\n",
    "____________________________________________________\n",
    "\n",
    "\n",
    "## weight towards individual variants\n",
    "\n",
    "```\n",
    "v1=20; v2=3; v3=3; v4=6; v5=1\n",
    "\n",
    "AUC Truth SNPs            :  0.6719735789816652\n",
    "AUC All SNPs in Variables :  0.6794628877442193\n",
    "AUC All SNPs in input file:  0.5286143098837526\n",
    "\n",
    "AUC Truth SNPs            :  0.6997608799617409\n",
    "AUC All SNPs in Variables :  0.6931490747193992\n",
    "AUC All SNPs in input file:  0.5773875449334499\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running VariantSpark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "#BASE=\"all-snps-header_10_8_6_4_2\"\n",
    "BASE=\"all-snps-header_5_4_3_2_1\"\n",
    "DATA1000GENOMES=\"subset.3.withTruth.vcf.bgz\" #3 for 0.5M snps and 2 for 5M snps\n",
    "S3DIR=\"s3://variant-spark/GigaScience/Data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: PEPS/Test/all-snps-header_5_4_3_2_1.pheno.csv to s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.pheno.csv\n",
      "upload: PEPS/Test/all-snps-header_5_4_3_2_1.varDataTruth.csv to s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.varDataTruth.csv\n",
      "upload: PEPS/Test/all-snps-header_5_4_3_2_1.TruthSNP.csv to s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.TruthSNP.csv\n"
     ]
    }
   ],
   "source": [
    "#upload the newly created files\n",
    "!aws s3 cp $LOCALDIR\"/\"$BASE\".pheno.csv\"      $S3DIR\"/PEPS/\"$BASE\".pheno.csv\"\n",
    "# also upload truth data for savekeeping\n",
    "!aws s3 cp $LOCALDIR\"/\"$BASE\".varDataTruth.csv\"      $S3DIR\"/PEPS/\"$BASE\".varDataTruth.csv\" \n",
    "!aws s3 cp $LOCALDIR\"/\"$BASE\".TruthSNP.csv\"      $S3DIR\"/PEPS/\"$BASE\".TruthSNP.csv\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20/01/20 04:52:24 INFO ImportanceCmd: Running with params: au.csiro.variantspark.cli.ImportanceCmd@73d62b5[featuresFile=s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.pheno.csv,featureColumn=lbl,outputFile=s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.vs.csv,nVariables=100000,includeData=true,randomSeed=13,modelFile=<null>,modelFileFormat=java,nTrees=500,rfMTry=-1,rfMTryFraction=0.1,rfEstimateOob=true,rfRandomizeEqual=false,rfSubsampleFraction=NaN,rfSampleNoReplacement=false,rfBatchSize=64,rfMaxDepth=2147483647,rfMinNodeSize=1,correctImportance=false,airRandomSeed=0,nOuputParitions=0,outputNormalization=raw,formats=org.json4s.DefaultFormats$@63917fe1,inputFile=s3://variant-spark/GigaScience/Data/1000Genomes/subset.3.withTruth.vcf.bgz,inputType=vcf,inputOptions=<null>,inputJSON=<null>,featureSource=<null>,beVerbose=false,beSilent=false,sparkPar=0,conf=org.apache.spark.SparkConf@720f56e2,spark=org.apache.spark.sql.SparkSession@677349fb,sc=org.apache.spark.SparkContext@4466cf5d,sqlContext=<null>]\n",
      "Finding  100000  most important features using random forest\n",
      "Loading header from VCF file: s3://variant-spark/GigaScience/Data/1000Genomes/subset.3.withTruth.vcf.bgz\n",
      "Loaded rows: [HG00096,HG00097,HG00099,...,NA21142,NA21143,NA21144] total: 2504\n",
      "Loaded variables: [1_15820_G_T,1_54165_G_C,1_55416_G_A,1_62017_A_G,1_62777_A_T,1_63268_T_C,...] total: 517729, took: 111.857\n",
      "Loading labels from: s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.pheno.csv, column: lbl\n",
      "20/01/20 04:54:16 INFO S3NativeFileSystem: Opening 's3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.pheno.csv' for reading\n",
      "Loaded labels: [1,0,1,...,0,0,1] total: 2504\n",
      "Training random forest with trees: 500 (batch size:  64)\n",
      "Random seed is: 13\n",
      "RF Params: au.csiro.variantspark.algo.RandomForestParams@17f7a1af[oob=true,nTryFraction=0.1,bootstrap=true,subsample=1.0,randomizeEquality=true,seed=13,maxDepth=2147483647,minNodeSize=1,correctImpurity=false,airRandomSeed=0]\n",
      "RF Params mTry: 51772\n",
      "Finished trees: 64, current oobError: 0.43570287539936103, totalTime: 281.761 s, avg timePerTree: 4.402515625 s\n",
      "Last build trees: 64, time: 281761 ms, timePerTree: 4402 ms\n",
      "Finished trees: 128, current oobError: 0.4193290734824281, totalTime: 575.857 s, avg timePerTree: 4.4988828125 s\n",
      "Last build trees: 64, time: 294096 ms, timePerTree: 4595 ms\n",
      "Finished trees: 192, current oobError: 0.420926517571885, totalTime: 869.11 s, avg timePerTree: 4.526614583333333 s\n",
      "Last build trees: 64, time: 293253 ms, timePerTree: 4582 ms\n",
      "Finished trees: 256, current oobError: 0.41333865814696485, totalTime: 1170.011 s, avg timePerTree: 4.57035546875 s\n",
      "Last build trees: 64, time: 300901 ms, timePerTree: 4701 ms\n",
      "Finished trees: 320, current oobError: 0.41293929712460065, totalTime: 1457.788 s, avg timePerTree: 4.5555875 s\n",
      "Last build trees: 64, time: 287777 ms, timePerTree: 4496 ms\n",
      "Finished trees: 384, current oobError: 0.4145367412140575, totalTime: 1751.504 s, avg timePerTree: 4.561208333333333 s\n",
      "Last build trees: 64, time: 293716 ms, timePerTree: 4589 ms\n",
      "Finished trees: 448, current oobError: 0.40934504792332266, totalTime: 2055.898 s, avg timePerTree: 4.589058035714285 s\n",
      "Last build trees: 64, time: 304394 ms, timePerTree: 4756 ms\n",
      "Finished trees: 500, current oobError: 0.41014376996805113, totalTime: 2298.625 s, avg timePerTree: 4.59725 s\n",
      "Last build trees: 52, time: 242727 ms, timePerTree: 4667 ms\n",
      "Random forest oob accuracy: 0.41014376996805113, took: 2298.827 s\n",
      "Variable importance preview\n",
      "21_19506651_G_A: 5.956927997557528\n",
      "9_79469334_T_C: 4.6590365117707675\n",
      "12_76362129_C_T: 0.6558799842524119\n",
      "12_20412031_A_T: 0.652729976413693\n",
      "7_71185237_G_A: 0.6455791846493515\n",
      "8_4075279_G_A: 0.6291522904675982\n",
      "5_11016111_G_T: 0.5387878273072324\n",
      "20/01/20 05:34:14 INFO MultipartUploadOutputStream: uploadPart: partNum 1 of 's3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.vs.csv' from local file '/mnt/s3/emrfs-4403151201893558791/0000000000', 134217728 bytes in 1837 ms, md5: sz4mLSXKv3640+oMNz1Fgw== md5hex: b33e262d25cabf7eb8d3ea0c373d4583\n",
      "20/01/20 05:35:16 INFO MultipartUploadOutputStream: uploadPart: partNum 2 of 's3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.vs.csv' from local file '/mnt1/s3/emrfs-2387768637491030874/0000000001', 134217728 bytes in 2071 ms, md5: wuvNKHzYRNJeLs8rcaLf4g== md5hex: c2ebcd287cd844d25e2ecf2b71a2dfe2\n",
      "20/01/20 05:35:36 INFO MultipartUploadOutputStream: close closed:false s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.vs.csv\n",
      "20/01/20 05:35:37 INFO MultipartUploadOutputStream: uploadPart: partNum 3 of 's3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.vs.csv' from local file '/mnt2/s3/emrfs-5128550870740535032/0000000002', 44682578 bytes in 687 ms, md5: OHI7fT4DQzbdDM6PHArItQ== md5hex: 38723b7d3e034336dd0cce8f1c0ac8b5\n",
      "20/01/20 05:35:37 INFO DefaultMultipartUploadDispatcher: Completed multipart upload of 3 parts 313118034 bytes\n"
     ]
    }
   ],
   "source": [
    "!biospark/bin/spark-submit --deploy-mode client --class au.csiro.variantspark.cli.VariantSparkApp \\\n",
    "/home/hadoop/biospark/lib/python3.6/site-packages/varspark/jars/variant-spark_2.11-0.3.0-SNAPSHOT-all.jar \\\n",
    "importance \\\n",
    "-if $S3DIR\"/1000Genomes/\"$DATA1000GENOMES \\\n",
    "-it vcf \\\n",
    "-ff $S3DIR\"/PEPS/\"$BASE\".pheno.csv\" \\\n",
    "-fc lbl \\\n",
    "-of $S3DIR\"/PEPS/\"$BASE\".vs.csv\" \\\n",
    "-on 100000 -rn 500 -rmtf 0.1 -rbs 100 -ro -ovn raw -sr 13 -od\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://variant-spark/GigaScience/Data/PEPS/all-snps-header_5_4_3_2_1.vs.csv to PEPS/Test/all-snps-header_5_4_3_2_1.vs.csv\n"
     ]
    }
   ],
   "source": [
    "### copy results back to master node\n",
    "!aws s3 cp $S3DIR\"/PEPS/\"$BASE\".vs.csv\" $LOCALDIR\"/\"$BASE\".vs.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "# when rerunning \n",
    "#!aws s3 cp $S3DIR\"/PEPS/\"$BASE\".pheno.csv\" $LOCALDIR\"/\"$BASE\".pheno.csv\"\n",
    "#!aws s3 cp $S3DIR\"/PEPS/\"$BASE\".TruthSNP.csv\" $LOCALDIR\"/\"$BASE\".TruthSNP.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>variable</th>\n",
       "      <th>21_19506651_G_A</th>\n",
       "      <th>9_79469334_T_C</th>\n",
       "      <th>12_76362129_C_T</th>\n",
       "      <th>12_20412031_A_T</th>\n",
       "      <th>7_71185237_G_A</th>\n",
       "      <th>8_4075279_G_A</th>\n",
       "      <th>5_11016111_G_T</th>\n",
       "      <th>3_60507811_G_T</th>\n",
       "      <th>17_35959746_G_A</th>\n",
       "      <th>1_53162156_C_A</th>\n",
       "      <th>...</th>\n",
       "      <th>2_76231009_C_A</th>\n",
       "      <th>15_37547903_C_A</th>\n",
       "      <th>4_60768061_A_C</th>\n",
       "      <th>4_119876232_G_T</th>\n",
       "      <th>15_40236372_G_A</th>\n",
       "      <th>19_34608021_A_T</th>\n",
       "      <th>7_22152629_A_G</th>\n",
       "      <th>1_86763477_T_C</th>\n",
       "      <th>8_37120185_G_A</th>\n",
       "      <th>7_20216348_C_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "variable  21_19506651_G_A  9_79469334_T_C  12_76362129_C_T  12_20412031_A_T  \\\n",
       "HG00096               2.0             2.0              2.0              0.0   \n",
       "HG00097               2.0             2.0              2.0              1.0   \n",
       "HG00099               2.0             1.0              1.0              0.0   \n",
       "HG00100               2.0             1.0              2.0              0.0   \n",
       "HG00101               1.0             2.0              1.0              0.0   \n",
       "\n",
       "variable  7_71185237_G_A  8_4075279_G_A  5_11016111_G_T  3_60507811_G_T  \\\n",
       "HG00096              0.0            2.0             0.0             0.0   \n",
       "HG00097              1.0            1.0             1.0             2.0   \n",
       "HG00099              1.0            0.0             1.0             2.0   \n",
       "HG00100              0.0            2.0             1.0             2.0   \n",
       "HG00101              1.0            2.0             2.0             1.0   \n",
       "\n",
       "variable  17_35959746_G_A  1_53162156_C_A       ...        2_76231009_C_A  \\\n",
       "HG00096               2.0             0.0       ...                   2.0   \n",
       "HG00097               2.0             2.0       ...                   0.0   \n",
       "HG00099               1.0             0.0       ...                   0.0   \n",
       "HG00100               0.0             0.0       ...                   1.0   \n",
       "HG00101               2.0             1.0       ...                   0.0   \n",
       "\n",
       "variable  15_37547903_C_A  4_60768061_A_C  4_119876232_G_T  15_40236372_G_A  \\\n",
       "HG00096               0.0             0.0              0.0              1.0   \n",
       "HG00097               2.0             1.0              0.0              0.0   \n",
       "HG00099               1.0             0.0              1.0              1.0   \n",
       "HG00100               1.0             1.0              0.0              0.0   \n",
       "HG00101               2.0             1.0              1.0              0.0   \n",
       "\n",
       "variable  19_34608021_A_T  7_22152629_A_G  1_86763477_T_C  8_37120185_G_A  \\\n",
       "HG00096               0.0             2.0             1.0             0.0   \n",
       "HG00097               0.0             0.0             1.0             0.0   \n",
       "HG00099               1.0             1.0             0.0             0.0   \n",
       "HG00100               0.0             2.0             0.0             1.0   \n",
       "HG00101               0.0             2.0             0.0             1.0   \n",
       "\n",
       "variable  7_20216348_C_T  \n",
       "HG00096              1.0  \n",
       "HG00097              1.0  \n",
       "HG00099              1.0  \n",
       "HG00100              2.0  \n",
       "HG00101              0.0  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sklearn\n",
    "\n",
    "# read the top 100 VariantSpark output \n",
    "FullVSoutput = pd.read_csv(LOCALDIR+\"/\"+BASE+\".vs.csv\") # full dataset\n",
    "\n",
    "snpData = FullVSoutput.copy(deep=True)\n",
    "snpData = snpData.set_index('variable').T\n",
    "snpData = snpData.drop('importance', axis=0)\n",
    "\n",
    "snpData100 = pd.DataFrame(FullVSoutput)[:100].copy(deep=True)\n",
    "snpData100 = snpData100.set_index('variable').T\n",
    "snpData100 = snpData100.drop('importance', axis=0)\n",
    "\n",
    "snpData100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>21_19506651_G_A</th>\n",
       "      <th>9_79469334_T_C</th>\n",
       "      <th>12_76362129_C_T</th>\n",
       "      <th>12_20412031_A_T</th>\n",
       "      <th>7_71185237_G_A</th>\n",
       "      <th>8_4075279_G_A</th>\n",
       "      <th>5_11016111_G_T</th>\n",
       "      <th>3_60507811_G_T</th>\n",
       "      <th>17_35959746_G_A</th>\n",
       "      <th>1_53162156_C_A</th>\n",
       "      <th>...</th>\n",
       "      <th>15_37547903_C_A</th>\n",
       "      <th>4_60768061_A_C</th>\n",
       "      <th>4_119876232_G_T</th>\n",
       "      <th>15_40236372_G_A</th>\n",
       "      <th>19_34608021_A_T</th>\n",
       "      <th>7_22152629_A_G</th>\n",
       "      <th>1_86763477_T_C</th>\n",
       "      <th>8_37120185_G_A</th>\n",
       "      <th>7_20216348_C_T</th>\n",
       "      <th>lbl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         21_19506651_G_A  9_79469334_T_C  12_76362129_C_T  12_20412031_A_T  \\\n",
       "HG00096              2.0             2.0              2.0              0.0   \n",
       "HG00097              2.0             2.0              2.0              1.0   \n",
       "HG00099              2.0             1.0              1.0              0.0   \n",
       "HG00100              2.0             1.0              2.0              0.0   \n",
       "HG00101              1.0             2.0              1.0              0.0   \n",
       "\n",
       "         7_71185237_G_A  8_4075279_G_A  5_11016111_G_T  3_60507811_G_T  \\\n",
       "HG00096             0.0            2.0             0.0             0.0   \n",
       "HG00097             1.0            1.0             1.0             2.0   \n",
       "HG00099             1.0            0.0             1.0             2.0   \n",
       "HG00100             0.0            2.0             1.0             2.0   \n",
       "HG00101             1.0            2.0             2.0             1.0   \n",
       "\n",
       "         17_35959746_G_A  1_53162156_C_A ...   15_37547903_C_A  \\\n",
       "HG00096              2.0             0.0 ...               0.0   \n",
       "HG00097              2.0             2.0 ...               2.0   \n",
       "HG00099              1.0             0.0 ...               1.0   \n",
       "HG00100              0.0             0.0 ...               1.0   \n",
       "HG00101              2.0             1.0 ...               2.0   \n",
       "\n",
       "         4_60768061_A_C  4_119876232_G_T  15_40236372_G_A  19_34608021_A_T  \\\n",
       "HG00096             0.0              0.0              1.0              0.0   \n",
       "HG00097             1.0              0.0              0.0              0.0   \n",
       "HG00099             0.0              1.0              1.0              1.0   \n",
       "HG00100             1.0              0.0              0.0              0.0   \n",
       "HG00101             1.0              1.0              0.0              0.0   \n",
       "\n",
       "         7_22152629_A_G  1_86763477_T_C  8_37120185_G_A  7_20216348_C_T  lbl  \n",
       "HG00096             2.0             1.0             0.0             1.0    1  \n",
       "HG00097             0.0             1.0             0.0             1.0    0  \n",
       "HG00099             1.0             0.0             0.0             1.0    1  \n",
       "HG00100             2.0             0.0             1.0             2.0    1  \n",
       "HG00101             2.0             0.0             1.0             0.0    1  \n",
       "\n",
       "[5 rows x 101 columns]"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the label created by PEPS and used during VS training and join to data\n",
    "label = pd.read_csv(LOCALDIR+\"/\"+BASE+\".pheno.csv\")\n",
    "label.head()\n",
    "jn=snpData.join(label.set_index('sample'))\n",
    "jn100=snpData100.join(label.set_index('sample'))\n",
    "jn100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>HG00103</th>\n",
       "      <th>HG00105</th>\n",
       "      <th>HG00106</th>\n",
       "      <th>...</th>\n",
       "      <th>NA21128</th>\n",
       "      <th>NA21129</th>\n",
       "      <th>NA21130</th>\n",
       "      <th>NA21133</th>\n",
       "      <th>NA21135</th>\n",
       "      <th>NA21137</th>\n",
       "      <th>NA21141</th>\n",
       "      <th>NA21142</th>\n",
       "      <th>NA21143</th>\n",
       "      <th>NA21144</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15820:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:62777:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:63268:T:C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:843405:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:1012483:G:T</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2505 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             SNP  HG00096  HG00097  HG00099  HG00100  HG00101  HG00102  \\\n",
       "0    1:15820:G:T        1        1        1        0        0        1   \n",
       "1    1:62777:A:T        1        1        0        1        1        2   \n",
       "2    1:63268:T:C        0        0        0        0        1        1   \n",
       "3   1:843405:A:G        1        1        0        0        2        1   \n",
       "4  1:1012483:G:T        0        2        2        2        2        2   \n",
       "\n",
       "   HG00103  HG00105  HG00106   ...     NA21128  NA21129  NA21130  NA21133  \\\n",
       "0        1        0        0   ...           1        0        0        1   \n",
       "1        0        1        1   ...           1        0        0        1   \n",
       "2        0        0        2   ...           1        0        0        2   \n",
       "3        0        1        2   ...           1        0        0        2   \n",
       "4        1        2        2   ...           1        1        1        2   \n",
       "\n",
       "   NA21135  NA21137  NA21141  NA21142  NA21143  NA21144  \n",
       "0        1        1        1        0        1        0  \n",
       "1        2        2        1        2        1        1  \n",
       "2        1        0        0        0        1        2  \n",
       "3        0        0        0        0        1        1  \n",
       "4        0        2        1        1        1        1  \n",
       "\n",
       "[5 rows x 2505 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw1000Genoms = pd.read_csv(PEPSINPUT+\".csv\")\n",
    "raw1000Genoms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10:128809541:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12:130740770:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12:131131186:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13:28892985:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14:49960242:A:G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNP\n",
       "0  10:128809541:A:G\n",
       "1  12:130740770:A:G\n",
       "2  12:131131186:C:T\n",
       "3   13:28892985:C:T\n",
       "4   14:49960242:A:G"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpDataTruthKeep = pd.read_csv(LOCALDIR+\"/\"+BASE+\".TruthSNP.csv\") # truth SNPs \n",
    "#snpDataTruthKeep['v'].replace({':': '_'},inplace=True, regex=True)\n",
    "snpDataTruthKeep.columns=['SNP']\n",
    "#snpDataTruthKeep.set_index('variable')\n",
    "#snpDataTruthKeep['value']=1\n",
    "snpDataTruthKeep.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17373, 2505)"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw1000Genoms.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28, 1)"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpDataTruthKeep.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "#raw1000Genoms.set_index('variable')\n",
    "#raw1000Genoms.sort_index(inplace=True)\n",
    "#snpDataTruthKeep.set_index('variable')\n",
    "#snpDataTruthKeep.sort_index(inplace=True)\n",
    "#snpDataTruthsub=raw1000Genoms.join(snpDataTruthKeep.set_index('SNP'))\n",
    "#snpDataTruthsub=raw1000Genoms.where(raw1000Genoms.variable==snpDataTruthKeep.variable).notna()\n",
    "#snpDataTruthsub.loc[snpDataTruthsub['value'] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>SNP</th>\n",
       "      <th>2_75815890_T_C</th>\n",
       "      <th>2_136827600_C_G</th>\n",
       "      <th>3_60507811_G_T</th>\n",
       "      <th>3_169355019_T_C</th>\n",
       "      <th>4_21622448_T_C</th>\n",
       "      <th>4_32192914_A_G</th>\n",
       "      <th>4_160997959_A_G</th>\n",
       "      <th>5_31869530_G_A</th>\n",
       "      <th>5_124084849_A_G</th>\n",
       "      <th>7_156114838_A_G</th>\n",
       "      <th>...</th>\n",
       "      <th>10_128809541_A_G</th>\n",
       "      <th>12_130740770_A_G</th>\n",
       "      <th>12_131131186_C_T</th>\n",
       "      <th>13_28892985_C_T</th>\n",
       "      <th>14_49960242_A_G</th>\n",
       "      <th>16_6495591_G_A</th>\n",
       "      <th>19_9114906_C_G</th>\n",
       "      <th>19_34608021_A_T</th>\n",
       "      <th>21_19506651_G_A</th>\n",
       "      <th>22_50067315_T_G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "SNP      2_75815890_T_C  2_136827600_C_G  3_60507811_G_T  3_169355019_T_C  \\\n",
       "HG00096               0                2               0                1   \n",
       "HG00097               1                0               2                2   \n",
       "HG00099               0                1               2                1   \n",
       "HG00100               0                0               2                0   \n",
       "HG00101               0                0               1                1   \n",
       "\n",
       "SNP      4_21622448_T_C  4_32192914_A_G  4_160997959_A_G  5_31869530_G_A  \\\n",
       "HG00096               1               1                2               2   \n",
       "HG00097               0               1                0               1   \n",
       "HG00099               2               2                1               2   \n",
       "HG00100               1               2                0               1   \n",
       "HG00101               0               2                0               0   \n",
       "\n",
       "SNP      5_124084849_A_G  7_156114838_A_G       ...         10_128809541_A_G  \\\n",
       "HG00096                2                0       ...                        1   \n",
       "HG00097                2                1       ...                        1   \n",
       "HG00099                2                1       ...                        2   \n",
       "HG00100                1                0       ...                        1   \n",
       "HG00101                1                0       ...                        0   \n",
       "\n",
       "SNP      12_130740770_A_G  12_131131186_C_T  13_28892985_C_T  14_49960242_A_G  \\\n",
       "HG00096                 2                 0                0                0   \n",
       "HG00097                 1                 0                1                0   \n",
       "HG00099                 2                 0                2                1   \n",
       "HG00100                 2                 0                0                1   \n",
       "HG00101                 2                 0                2                1   \n",
       "\n",
       "SNP      16_6495591_G_A  19_9114906_C_G  19_34608021_A_T  21_19506651_G_A  \\\n",
       "HG00096               0               1                0                2   \n",
       "HG00097               1               1                0                2   \n",
       "HG00099               1               0                1                2   \n",
       "HG00100               1               0                0                2   \n",
       "HG00101               2               0                0                1   \n",
       "\n",
       "SNP      22_50067315_T_G  \n",
       "HG00096                1  \n",
       "HG00097                0  \n",
       "HG00099                1  \n",
       "HG00100                1  \n",
       "HG00101                0  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#snpDataTruth.sort_index(inplace=True)\n",
    "#snpDataTruthKeep.sort_index(inplace=True)\n",
    "#snpDataTruthsub=snpDataTruth.loc[snpDataTruth['variable'].isin(snpDataTruthKeep)]\n",
    "#snpDataTruthsub=raw1000Genoms.where(raw1000Genoms.variable==snpDataTruthKeep.variable).notna()\n",
    "snpDataTruthsub = pd.merge(raw1000Genoms,snpDataTruthKeep, on=['SNP'], how='inner')\n",
    "#snpDataTruthsub=raw1000Genoms.join(snpDataTruthKeep.set_index('variable'))\n",
    "#snpDataTruthsub=snpDataTruthsub.loc[snpDataTruthsub['k'] == 1]\n",
    "snpDataTruthsub['SNP'].replace({':': '_'},inplace=True, regex=True)\n",
    "snpDataTruthsub=snpDataTruthsub.set_index('SNP').T\n",
    "#snpDataTruthsub = snpDataTruthsub.drop('importance', axis=0)\n",
    "\n",
    "#snpDataTruthsub.shape\n",
    "snpDataTruthsub.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2_75815890_T_C</th>\n",
       "      <th>2_136827600_C_G</th>\n",
       "      <th>3_60507811_G_T</th>\n",
       "      <th>3_169355019_T_C</th>\n",
       "      <th>4_21622448_T_C</th>\n",
       "      <th>4_32192914_A_G</th>\n",
       "      <th>4_160997959_A_G</th>\n",
       "      <th>5_31869530_G_A</th>\n",
       "      <th>5_124084849_A_G</th>\n",
       "      <th>7_156114838_A_G</th>\n",
       "      <th>...</th>\n",
       "      <th>12_130740770_A_G</th>\n",
       "      <th>12_131131186_C_T</th>\n",
       "      <th>13_28892985_C_T</th>\n",
       "      <th>14_49960242_A_G</th>\n",
       "      <th>16_6495591_G_A</th>\n",
       "      <th>19_9114906_C_G</th>\n",
       "      <th>19_34608021_A_T</th>\n",
       "      <th>21_19506651_G_A</th>\n",
       "      <th>22_50067315_T_G</th>\n",
       "      <th>lbl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         2_75815890_T_C  2_136827600_C_G  3_60507811_G_T  3_169355019_T_C  \\\n",
       "HG00096               0                2               0                1   \n",
       "HG00097               1                0               2                2   \n",
       "HG00099               0                1               2                1   \n",
       "HG00100               0                0               2                0   \n",
       "HG00101               0                0               1                1   \n",
       "\n",
       "         4_21622448_T_C  4_32192914_A_G  4_160997959_A_G  5_31869530_G_A  \\\n",
       "HG00096               1               1                2               2   \n",
       "HG00097               0               1                0               1   \n",
       "HG00099               2               2                1               2   \n",
       "HG00100               1               2                0               1   \n",
       "HG00101               0               2                0               0   \n",
       "\n",
       "         5_124084849_A_G  7_156114838_A_G ...   12_130740770_A_G  \\\n",
       "HG00096                2                0 ...                  2   \n",
       "HG00097                2                1 ...                  1   \n",
       "HG00099                2                1 ...                  2   \n",
       "HG00100                1                0 ...                  2   \n",
       "HG00101                1                0 ...                  2   \n",
       "\n",
       "         12_131131186_C_T  13_28892985_C_T  14_49960242_A_G  16_6495591_G_A  \\\n",
       "HG00096                 0                0                0               0   \n",
       "HG00097                 0                1                0               1   \n",
       "HG00099                 0                2                1               1   \n",
       "HG00100                 0                0                1               1   \n",
       "HG00101                 0                2                1               2   \n",
       "\n",
       "         19_9114906_C_G  19_34608021_A_T  21_19506651_G_A  22_50067315_T_G  \\\n",
       "HG00096               1                0                2                1   \n",
       "HG00097               1                0                2                0   \n",
       "HG00099               0                1                2                1   \n",
       "HG00100               0                0                2                1   \n",
       "HG00101               0                0                1                0   \n",
       "\n",
       "         lbl  \n",
       "HG00096    1  \n",
       "HG00097    0  \n",
       "HG00099    1  \n",
       "HG00100    1  \n",
       "HG00101    1  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the label created by PEPS and used during VS training and join to data\n",
    "jnAll=snpDataTruthsub.join(label.set_index('sample'))\n",
    "jnAll.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "#snpDataTruth = snpDataTruth.set_index('variable').T\n",
    "#snpDataTruth = snpDataTruth.drop('importance', axis=0)\n",
    "#snpDataTruth.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>21_19506651_G_A</th>\n",
       "      <th>9_79469334_T_C</th>\n",
       "      <th>12_76362129_C_T</th>\n",
       "      <th>12_20412031_A_T</th>\n",
       "      <th>7_71185237_G_A</th>\n",
       "      <th>8_4075279_G_A</th>\n",
       "      <th>5_11016111_G_T</th>\n",
       "      <th>3_60507811_G_T</th>\n",
       "      <th>17_35959746_G_A</th>\n",
       "      <th>1_53162156_C_A</th>\n",
       "      <th>...</th>\n",
       "      <th>15_37547903_C_A</th>\n",
       "      <th>4_60768061_A_C</th>\n",
       "      <th>4_119876232_G_T</th>\n",
       "      <th>15_40236372_G_A</th>\n",
       "      <th>19_34608021_A_T</th>\n",
       "      <th>7_22152629_A_G</th>\n",
       "      <th>1_86763477_T_C</th>\n",
       "      <th>8_37120185_G_A</th>\n",
       "      <th>7_20216348_C_T</th>\n",
       "      <th>lbl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         21_19506651_G_A  9_79469334_T_C  12_76362129_C_T  12_20412031_A_T  \\\n",
       "HG00096              2.0             2.0              2.0              0.0   \n",
       "HG00097              2.0             2.0              2.0              1.0   \n",
       "HG00099              2.0             1.0              1.0              0.0   \n",
       "HG00100              2.0             1.0              2.0              0.0   \n",
       "HG00101              1.0             2.0              1.0              0.0   \n",
       "\n",
       "         7_71185237_G_A  8_4075279_G_A  5_11016111_G_T  3_60507811_G_T  \\\n",
       "HG00096             0.0            2.0             0.0             0.0   \n",
       "HG00097             1.0            1.0             1.0             2.0   \n",
       "HG00099             1.0            0.0             1.0             2.0   \n",
       "HG00100             0.0            2.0             1.0             2.0   \n",
       "HG00101             1.0            2.0             2.0             1.0   \n",
       "\n",
       "         17_35959746_G_A  1_53162156_C_A ...   15_37547903_C_A  \\\n",
       "HG00096              2.0             0.0 ...               0.0   \n",
       "HG00097              2.0             2.0 ...               2.0   \n",
       "HG00099              1.0             0.0 ...               1.0   \n",
       "HG00100              0.0             0.0 ...               1.0   \n",
       "HG00101              2.0             1.0 ...               2.0   \n",
       "\n",
       "         4_60768061_A_C  4_119876232_G_T  15_40236372_G_A  19_34608021_A_T  \\\n",
       "HG00096             0.0              0.0              1.0              0.0   \n",
       "HG00097             1.0              0.0              0.0              0.0   \n",
       "HG00099             0.0              1.0              1.0              1.0   \n",
       "HG00100             1.0              0.0              0.0              0.0   \n",
       "HG00101             1.0              1.0              0.0              0.0   \n",
       "\n",
       "         7_22152629_A_G  1_86763477_T_C  8_37120185_G_A  7_20216348_C_T  lbl  \n",
       "HG00096             2.0             1.0             0.0             1.0    1  \n",
       "HG00097             0.0             1.0             0.0             1.0    0  \n",
       "HG00099             1.0             0.0             0.0             1.0    1  \n",
       "HG00100             2.0             0.0             1.0             2.0    1  \n",
       "HG00101             2.0             0.0             1.0             0.0    1  \n",
       "\n",
       "[5 rows x 101 columns]"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jn100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "#helper function to train dataset\n",
    "\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import roc_auc_score\n",
    "\n",
    "def RF_AUC(dfx, nTree):\n",
    "    df = dfx.copy()\n",
    "    features = df.columns[:-1]\n",
    "    \n",
    "    df['is_train'] = np.random.uniform(0, 1, len(df)) <= .75\n",
    "    train, test = df[df['is_train']==True], df[df['is_train']==False]\n",
    "    clf = RandomForestClassifier(n_jobs=2, n_estimators=nTree, random_state=0)\n",
    "    clf.fit(train[features], train['lbl'])\n",
    "    #clf.predict(test[features])\n",
    "    prob = clf.predict_proba(test[features])\n",
    "    y_true = test['lbl']\n",
    "    y_scores = prob[:,1]\n",
    "    return clf, roc_auc_score(y_true, y_scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUC Total                    :  0.7427731431332678 sem 0.007674976688547618\n",
      "AUC VariantSpark 100         :  0.7003592726668074 sem 0.004613436579131516\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import sem\n",
    "from statistics import mean \n",
    "\n",
    "numTree=500\n",
    "\n",
    "AucTruthAr=[]\n",
    "AucTruthAr100=[]\n",
    "\n",
    "# run 5 times to get a feel for the variation\n",
    "for i in range(0,5):\n",
    "    clfTruth, AucTruth = RF_AUC(jnAll, nTree=numTree)\n",
    "    AucTruthAr.append(AucTruth)\n",
    "    clfTruth, AucTruth = RF_AUC(jn100, nTree=numTree)\n",
    "    AucTruthAr100.append(AucTruth)\n",
    "\n",
    "print(\"AUC Total                    : \", mean(AucTruthAr),\"sem\", sem(AucTruthAr) )\n",
    "print(\"AUC VariantSpark 100         : \", mean(AucTruthAr100),\"sem\", sem(AucTruthAr100) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VS found  9  out of  28 truth variables\n"
     ]
    }
   ],
   "source": [
    "snpDataTruthKeep.columns=['variable']\n",
    "snpDataTruthKeep['truth']=1\n",
    "snpDataTruthKeep['variable'].replace({':': '_'},inplace=True, regex=True)\n",
    "VScorrect = pd.merge(FullVSoutput[:100],snpDataTruthKeep, on=['variable'], how='inner')\n",
    "\n",
    "print (\"VS found \", VScorrect.shape[0],\" out of \", snpDataTruthKeep.shape[0],\"truth variables\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variable</th>\n",
       "      <th>importance</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>HG00103</th>\n",
       "      <th>HG00105</th>\n",
       "      <th>...</th>\n",
       "      <th>NA21129</th>\n",
       "      <th>NA21130</th>\n",
       "      <th>NA21133</th>\n",
       "      <th>NA21135</th>\n",
       "      <th>NA21137</th>\n",
       "      <th>NA21141</th>\n",
       "      <th>NA21142</th>\n",
       "      <th>NA21143</th>\n",
       "      <th>NA21144</th>\n",
       "      <th>truth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21_19506651_G_A</td>\n",
       "      <td>5.956928</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9_79469334_T_C</td>\n",
       "      <td>4.659037</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12_76362129_C_T</td>\n",
       "      <td>0.655880</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12_20412031_A_T</td>\n",
       "      <td>0.652730</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7_71185237_G_A</td>\n",
       "      <td>0.645579</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2507 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          variable  importance  HG00096  HG00097  HG00099  HG00100  HG00101  \\\n",
       "0  21_19506651_G_A    5.956928        2        2        2        2        1   \n",
       "1   9_79469334_T_C    4.659037        2        2        1        1        2   \n",
       "2  12_76362129_C_T    0.655880        2        2        1        2        1   \n",
       "3  12_20412031_A_T    0.652730        0        1        0        0        0   \n",
       "4   7_71185237_G_A    0.645579        0        1        1        0        1   \n",
       "\n",
       "   HG00102  HG00103  HG00105  ...    NA21129  NA21130  NA21133  NA21135  \\\n",
       "0        1        1        1  ...          1        1        1        0   \n",
       "1        2        1        2  ...          2        2        2        2   \n",
       "2        1        1        2  ...          0        0        0        1   \n",
       "3        0        0        0  ...          2        1        2        1   \n",
       "4        1        1        2  ...          1        1        1        0   \n",
       "\n",
       "   NA21137  NA21141  NA21142  NA21143  NA21144  truth  \n",
       "0        0        2        1        0        1    1.0  \n",
       "1        2        2        2        2        2    1.0  \n",
       "2        1        0        1        0        1    0.0  \n",
       "3        0        2        1        0        0    0.0  \n",
       "4        0        0        1        0        1    0.0  \n",
       "\n",
       "[5 rows x 2507 columns]"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VScorrect = pd.merge(FullVSoutput,snpDataTruthKeep, on=['variable'], how='left')\n",
    "VScorrect=VScorrect.fillna(0)\n",
    "VScorrect.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MannwhitneyuResult(statistic=882.0, pvalue=2.0241158356994545e-26)"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "stats.mannwhitneyu(VScorrect.importance[:100], VScorrect.truth[:100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "# append rbo path to syspath\n",
    "sys.path.append('/home/hadoop/src/rbo')\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variable</th>\n",
       "      <th>importance</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>HG00103</th>\n",
       "      <th>HG00105</th>\n",
       "      <th>...</th>\n",
       "      <th>NA21129</th>\n",
       "      <th>NA21130</th>\n",
       "      <th>NA21133</th>\n",
       "      <th>NA21135</th>\n",
       "      <th>NA21137</th>\n",
       "      <th>NA21141</th>\n",
       "      <th>NA21142</th>\n",
       "      <th>NA21143</th>\n",
       "      <th>NA21144</th>\n",
       "      <th>truth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21_19506651_G_A</td>\n",
       "      <td>5.956928</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9_79469334_T_C</td>\n",
       "      <td>4.659037</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12_76362129_C_T</td>\n",
       "      <td>0.655880</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12_20412031_A_T</td>\n",
       "      <td>0.652730</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7_71185237_G_A</td>\n",
       "      <td>0.645579</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2507 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          variable  importance  HG00096  HG00097  HG00099  HG00100  HG00101  \\\n",
       "0  21_19506651_G_A    5.956928        2        2        2        2        1   \n",
       "1   9_79469334_T_C    4.659037        2        2        1        1        2   \n",
       "2  12_76362129_C_T    0.655880        2        2        1        2        1   \n",
       "3  12_20412031_A_T    0.652730        0        1        0        0        0   \n",
       "4   7_71185237_G_A    0.645579        0        1        1        0        1   \n",
       "\n",
       "   HG00102  HG00103  HG00105  ...    NA21129  NA21130  NA21133  NA21135  \\\n",
       "0        1        1        1  ...          1        1        1        0   \n",
       "1        2        1        2  ...          2        2        2        2   \n",
       "2        1        1        2  ...          0        0        0        1   \n",
       "3        0        0        0  ...          2        1        2        1   \n",
       "4        1        1        2  ...          1        1        1        0   \n",
       "\n",
       "   NA21137  NA21141  NA21142  NA21143  NA21144  truth  \n",
       "0        0        2        1        0        1    1.0  \n",
       "1        2        2        2        2        2    1.0  \n",
       "2        1        0        1        0        1    0.0  \n",
       "3        0        2        1        0        0    0.0  \n",
       "4        0        0        1        0        1    0.0  \n",
       "\n",
       "[5 rows x 2507 columns]"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VScorrect.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variable</th>\n",
       "      <th>truth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10_128809541_A_G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12_130740770_A_G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12_131131186_C_T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13_28892985_C_T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14_49960242_A_G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           variable  truth\n",
       "0  10_128809541_A_G      1\n",
       "1  12_130740770_A_G      1\n",
       "2  12_131131186_C_T      1\n",
       "3   13_28892985_C_T      1\n",
       "4   14_49960242_A_G      1"
      ]
     },
     "execution_count": 275,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpDataTruthKeep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rank-biased-overlap  0.0889588920296764\n"
     ]
    }
   ],
   "source": [
    "import rbo\n",
    "\n",
    "#S = [1, 2, 3]; T = [1, 3, 2]\n",
    "#rbo.RankingSimilarity(S, T).rbo()\n",
    "\n",
    "v=rbo.RankingSimilarity(VScorrect.variable[:100].values, snpDataTruthKeep.variable.values).rbo()\n",
    "print(\"Rank-biased-overlap \",v)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results\n",
    "\n",
    "```\n",
    "v1=10; v2=8; v3=6; v4=4; v5=2\n",
    "AUC Total                    :  0.7186311759007349 sem 0.012392640730408109\n",
    "AUC VariantSpark 100         :  0.7400257375096144 sem 0.016097212737146947\n",
    "\n",
    "VS found  11  out of  57 truth variables\n",
    "MannwhitneyuResult(statistic=1078.0, pvalue=5.352942586230578e-24)\n",
    "\n",
    "\n",
    "\n",
    "v1=5; v2=4; v3=3; v4=2; v5=1\n",
    "\n",
    "AUC Total                    :  0.7427731431332678 sem 0.007674976688547618\n",
    "AUC VariantSpark 100         :  0.7003592726668074 sem 0.004613436579131516\n",
    "\n",
    "VS found  9  out of  28 truth variables\n",
    "\n",
    "MannwhitneyuResult(statistic=882.0, pvalue=2.0241158356994545e-26)\n",
    "Rank-biased-overlap  0.0889588920296764\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# +++++++++++++ Dont need the rest ++++++++++++++++"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Transform VS csv to VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using variant-spark jar at '/home/hadoop/biospark/lib/python3.6/site-packages/varspark/jars/variant-spark_2.11-0.3.0-SNAPSHOT-all.jar'\n",
      "using hail jar at '/home/hadoop/biospark/lib/python3.6/site-packages/hail/hail-all-spark.jar'\n",
      "using hail jar at /home/hadoop/biospark/lib/python3.6/site-packages/hail/hail-all-spark.jar\n",
      "Running on Apache Spark version 2.4.4\n",
      "SparkUI available at http://ip-172-31-35-120.ap-southeast-2.compute.internal:4040\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.16-6da0d3571629\n",
      "LOGGING: writing to /home/hadoop/hail-20200113-0304-0.2.16-6da0d3571629.log\n"
     ]
    }
   ],
   "source": [
    "from pyspark import SparkContext\n",
    "sc = SparkContext()\n",
    "import hail as hl\n",
    "import varspark.hail as vshl\n",
    "vshl.init(sc=sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t7565677\t7565678\n",
      "1\t18535254\t18535255\n",
      "1\t22577936\t22577937\n",
      "1\t92140325\t92140326\n",
      "1\t113349346\t113349347\n",
      "1\t239574660\t239574661\n",
      "2\t1625090\t1625091\n",
      "2\t3110757\t3110758\n",
      "2\t53567632\t53567633\n",
      "2\t71300515\t71300516\n",
      "100 /home/hadoop/input-small_VS.bed\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "## get the top 100 elements and transform to bed\n",
    "tail -n +2 /home/hadoop/input-small_VS.csv  | head -n 100 \\\n",
    "    | gawk '{split($0,a,\"_\"); print a[1]\"\\t\"a[2]-1\"\\t\"a[2]}' | \\\n",
    "    sort -k1n -k2n > /home/hadoop/input-small_VS.bed\n",
    "head /home/hadoop/input-small_VS.bed\n",
    "wc -l /home/hadoop/input-small_VS.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Completed 2.0 KiB/2.0 KiB (21.1 KiB/s) with 1 file(s) remaining\r",
      "upload: ./input-small_VS.bed to s3://csiro-tb/home/bau04c/test/input-small_VS.bed\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "#uload bed file to S3\n",
    "aws s3 cp /home/hadoop/input-small_VS.bed            s3://csiro-tb/home/bau04c/test/input-small_VS.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-01-13 03:36:35 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(162, 2504)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading the original vcf file into Hail\n",
    "data = hl.import_vcf('s3://csiro-tb/home/bau04c/test/test.vcf',reference_genome='GRCh37')\n",
    "data.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-01-13 03:36:51 Hail: INFO: Reading table with no type imputation\n",
      "  Loading column 'f0' as type 'str' (user-specified)\n",
      "  Loading column 'f1' as type 'int32' (user-specified)\n",
      "  Loading column 'f2' as type 'int32' (user-specified)\n",
      "\n",
      "2020-01-13 03:36:52 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(100, 2504)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interval_table = hl.import_bed('s3://csiro-tb/home/bau04c/test/input-small_VS.bed', reference_genome='GRCh37')\n",
    "filtered_data = data.filter_rows(hl.is_defined(interval_table[data.locus]))\n",
    "filtered_data.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-01-13 03:37:07 Hail: INFO: Coerced sorted dataset\n",
      "2020-01-13 03:37:08 Hail: INFO: merging 2 files totalling 1007.7K...\n",
      "2020-01-13 03:37:08 Hail: INFO: while writing:\n",
      "    s3://csiro-tb/home/bau04c/test/test_VS.vcf\n",
      "  merge time: 127.413ms\n"
     ]
    }
   ],
   "source": [
    "hl.export_vcf(filtered_data, 's3://csiro-tb/home/bau04c/test/test_VS.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Completed 256.0 KiB/1007.7 KiB (3.0 MiB/s) with 1 file(s) remaining\r",
      "Completed 512.0 KiB/1007.7 KiB (5.6 MiB/s) with 1 file(s) remaining\r",
      "Completed 768.0 KiB/1007.7 KiB (7.8 MiB/s) with 1 file(s) remaining\r",
      "Completed 1007.7 KiB/1007.7 KiB (9.8 MiB/s) with 1 file(s) remaining\r",
      "download: s3://csiro-tb/home/bau04c/test/test_VS.vcf to ./input-small_VS.vcf\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "### to copy results back to master node\n",
    "aws s3 cp s3://csiro-tb/home/bau04c/test/test_VS.vcf /home/hadoop/input-small_VS.vcf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
